src/vehicle.cpp

```diff
@@ -3504,35 +3383,124 @@ int vehicle::charge_battery (int amount, bool include_other_vehicles)
+                    int qty = std::min( amount, int( pt.ammo_remaining() * double( density ) / 1000 ) );
+
+                    // calculate reactor charges that will be consumed...
+                    double burn = double( qty ) / density;
+
+                    // ...partial charges have a proportional chance of being consumed each turn
+                    burn += x_in_y( fmod( burn, 1.0 ) * 1000, 1000 ) ? 1 : 0;
+
+                    pt.ammo_consume( burn, global_part_pos3( pt ) );
+                    amount -= qty;
+
+                    if( amount == 0 ) {
+                        break;
+                    }
+                }
+            }
+        }
     }
 
-    return amount; // non-zero if we weren't able to fulfill demand.
+    return amount;
 }
 
 void vehicle::idle(bool on_map) {
     auto &eng = current_engine();
+
+    // power usage is in watts but battery storage (kJ) is less granular...
+    int deficit = power_usage();
+
+    // ...so partial charges have a proportional chance of being consumed each turn
+    deficit += x_in_y( deficit % 1000, 1000 ) ? 1000 : 0;
+
+    // electrical power not supplied must be found from alternators
+    deficit = discharge( deficit / 1000, true, true );
+
+    // calculate maximum possible output from all active alternators
+    int generate = 0;
     if( engine_on && eng ) {
+        for( const auto &e : parts ) {
+            if( e.is_alternator() && !e.is_broken() && e.mount == eng.mount ) {
+                generate += e.info().epower;
+            }
+        }
+    }
+
+    if( deficit > generate ) {
+        // insufficient power so disable electrical parts
+        for( auto &e : parts ) {
+            if( e.enabled && e.info().epower < 0 ) {
+                e.enabled = false;
+            }
+        }
+
+        // handle parts using legacy global state
+        is_alarm_on = false;
+        camera_on = false;
+
+        if( player_in_control( g->u ) || g->u.sees( global_pos3() ) ) {
+            add_msg( _( "The %s's battery dies!" ), name.c_str() );
+        }
+
+        if( eng.info().epower < 0 ) {
+            engine_on = false;
+            if( player_in_control( g->u ) || g->u.sees( global_pos3() ) ) {
+                add_msg( _( "The %s's engine dies!" ), name.c_str() );
+            }
+        }
+
+    } else {
+        // as above power generation is in watts but battery storage (kJ) is less granular...
+        int recharge = generate - deficit;
+
+        // ...so partial charges have a proportional chance of being added each turn
+        recharge += x_in_y( recharge % 1000, 1000 ) ? 1000 : 0;
+
+        // attempt to recharge battery limiting generated power to total battery capacity
+        generate -= charge_battery( recharge );
+    }
+
+    if( engine_on && eng ) {
+        // calculate load on engine from both propulsion and alternators
+        int load = friction_load() + generate;
```

Comment:
That's engine efficiency, but no generator efficiency. Are they all assumed to be near-100% efficient?